채팅 기능 보완

devjaewon opened the pull request:
# 채팅 기능 보완

## 해당 이슈 📎

- [Epic#15, Epic#41] 채팅 기능 보완 (#55)

## 변경 사항 🛠

- 채팅이 쭉 쌓일 때, 스크롤바의 상태가 가장 최근 채팅을 볼 수 있도록 맞춰져야 한다.
- 처음 화면 진입시 기존의 채팅들을 불러와야 한다.
- 채팅을 하면 인풋창이 초기화 되어야 한다.
- 채팅창에서 엔터키를 이용해도 전송이 가능하다.

## 테스트 ✨

> 없음

## 리뷰어 참고 사항 🙋‍♀️

> 없음

jjazan committed:
[ FEATURE ] 채팅 전송시 Input 클리어 및 엔터키로 전송 (#55)

Why
- UX 측면에서 엔터키를 이용해 전송이 가능해야한다.
- 전송이 되었다면 채팅창이 클리어되어야한다.

What
- ChatInput 컴포넌트에서 채팅 전송시 message state 초기화
- ChatInput 컴포넌트에서 Enter Key 에 대한 이벤트 추가
- ChatLogs 에서 엔터키 적용 가능하도록 수정

jjazan committed:
[ FEATURE ] 채팅 로그를 가져오는 API 완성 (#55)

Why
- 페이지 진입시 기존의 채팅을 모두 가져와야한다.

What
- getChatLogs Query 추가
- model 을 payload 로 변경하는 부분에 대해 메서드 생성

jjazan committed:
[ FEATURE ] 채팅 로그를 가져오는 Apollo Cache 기능 완성 (#55)

Why
- 페이지 진입시 기존의 채팅을 모두 가져와야한다.

What
- 페이지 진입시 채팅로그를 모두 가져와 캐시에 저장하고, broadcast 한다.
- broadcast 된 쿼리를 받기 위해서, readQuery 대신 useQuery 를 활용한다.

jjazan committed:
[ FEATURE ] 채팅시 채팅로그 UI가 가장 최근의 채팅을 따라가도록 한다. (#55)

Why
- 채팅은 커뮤니케이션을 위한 것으로 실시간 서로의 의견을 볼 수 있어야한다.
- 따라서 최근의 채팅이 계속 보이는 것이 적절한 UI/UX 이다

What
- 스크롤 마지막 부분을 계산하는 dom Util 추가
- 채팅 로그가 올라올 때, 스크롤 마지막으로 이동

happydhKim commented on server-api/src/express/app.js:
```diff
@@ -9,7 +9,7 @@ const corsOption = {
   methods: 'GET,HEAD,PUT,PATCH,POST,DELETE',
   credentials: true,
-  exposedHeaders: ['x-auth-token'],
+  exposedHeaders: ['x-auth-token', 'x-anonymous-token'],
```
와우 드디어 로그인 안한 사용자도 따봉을 할 수 있겠네요 멋집니다 🌝

devjaewon replied:
굳굳~ 😄

happydhKim commented on web/src/components/channel/Chat/ChatLogs/style.jsx:
```diff
@@ -0,0 +1,34 @@
+    overflow-x: hidden;
+    overflow-y: auto;
+    padding: 0 ${px(20)};
+    scrollbar-width: none; /* For Firefox */
```
와우 다른 브라우저까지 고려를 해주셨네요!!
다음 회의 때 논의를 해서 다 같이 하는게 좋을거같아요!
꼭 말씀해주세요🥺

devjaewon replied:
네 알겠습니다~! 크로스 브라우징 관련 논의를 한번 가질까요!

happydhKim reviewed (approved):
참 잘하셨어요!!
코드가 고급스럽네요!
스타일 부분에 대해서 회의를 했으면 합니다.( px, ... )
드로피가 성장하는게 보여서 좋습니다.👨‍👨‍👧‍👧

aereeeee commented on server-api/src/models/chats.js:
```diff
@@ -29,4 +30,21 @@ const ChatSchema = new Schema({
   },
 });
 
+ChatSchema.methods.toPayload = function toChatPayload(...objs) {
```
graphQL을 위한 payload 스키마를 따로 빼니 훨씬 코드가 깔끔해지네요! 멋집니다:)

devjaewon replied:
굳굳~👍

aereeeee commented on web/src/components/channel/Chat/ChatInput/ChatInput.jsx:
```diff
@@ -3,27 +3,40 @@ import PropTypes from 'prop-types';
+    setMessage('');
+  };
+  const handleChangeInput = (event) => {
+    if (event);
```
체크해주세용~~!

devjaewon replied:
헉 실수했네요.. 넵!!

aereeeee commented on web/src/components/channel/Chat/ChatInput/ChatInput.jsx:
```diff
@@ -3,27 +3,40 @@ import PropTypes from 'prop-types';
-        placeholder="질문 또는 의견을 남겨주세요."
-        onChange={handleChange}
-        value={state}
+        placeholder="의견을 남겨주세요."
```
placeholder의 내용은 나중에 한번에 관리할 수 있게 constants로 빼면 어떨까요?

devjaewon replied:
constants 로 관리하는 것 좋습니다~! placeholder constants 추가로 만들겠습니다!

always-awake commented on server-api/src/apollo/resolvers/chats.js:
```diff
@@ -12,14 +11,7 @@ const addChat = async (_, { channelId, message }, { user, pubsub }) => {
-      likes: [],
-      createdAt: newChat.createdAt,
-    };
+    const payload = newChat.toPayload({ author: user });
```
모델의 메소드로 toPayload를 정의하고, 사용하시니까 중복이 정말 많이 준 것 같습니다! 👍

aereeeee commented on web/src/hooks/useInitChatLogs.js:
```diff
@@ -0,0 +1,44 @@
+    };
+
+    client.writeQuery({ query: GET_CHAT_CACHED, data });
+  }, [result.loading]);
```
캐시에 데이터를 쓰는 로직이 전에는 reslover폴더에 있었는데 컴포넌트로 빼신 이유가 있나요?

devjaewon replied:
해당 파일 내에서 readQuery 를 하고 있기 때문에, resolver에 넣고 `useMutation`을 활용하기 보다 `writeQuery`를 바로 해주는 것이 가독성 측면에서 더 좋다고 생각했습니다.

aereeeee reviewed (commented):
멋진 코드 잘 봤습니다 :-) 키보드가 또 말썽이라 이모티콘을 달 수 없어 슬프네요 thumb이모티콘을 달고싶은 코드입니다:)

always-awake reviewed (approved):
수고 많으셨습니당 전체적으로 코드가 깔끔해서 좋았습니당 😃

jjazan committed:
[ REFACTOR ] 리뷰사항 반영 (#55)

Why
- 스타일드 컴포넌트 네이밍 방식을 통일할 필요가 있다.
- 불필요한 코드가 삽입되어 있다.

What
- if (event) 삭제
- 컴포넌트 이름과 최상위 스타일드 컴포넌트 이름을 통일시킨다.

aereeeee committed:
Merge branch 'develop' into feature/issue55
